signer/src/bitcoin/validation.rs

```diff
@@ -28,129 +37,396 @@ pub struct BitcoinTxContext {
+            tx_fee: Amount::from_sat(tx.tx_fee),
+            reports,
+            chain_tip_height: self.chain_tip_height,
+        };
+
+        Ok((out, signer_state))
+    }
+}
+
+/// An intermediate struct to aid in computing validation of deposits and
+/// withdrawals and transforming the computed sighash into a
+/// [`BitcoinSighash`].
+pub struct BitcoinTxValidationData {
+    /// The sighash of the signers' prevout
+    pub signer_sighash: SignatureHash,
+    /// The sighash of each of the deposit request prevout
+    pub deposit_sighashes: Vec<SignatureHash>,
+    /// The computed deposits and withdrawals reports.
+    pub reports: SbtcReports,
+    /// The chain tip at the time that this signer received the sign
+    /// request.
+    pub chain_tip: BitcoinBlockHash,
+    /// The transaction that we are (implicitly) requested to help sign.
+    pub tx: bitcoin::Transaction,
+    /// the transaction fee in sats
+    pub tx_fee: Amount,
+    /// the chain tip height.
+    pub chain_tip_height: u64,
+}
+
+/// The version of the algorithm that constructed the bitcoin transaction.
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, PartialOrd, Ord, sqlx::Type, strum::Display)]
+#[sqlx(type_name = "varchar", rename_all = "snake_case")]
+#[derive(serde::Serialize, serde::Deserialize)]
+#[strum(serialize_all = "snake_case")]
+#[cfg_attr(feature = "testing", derive(fake::Dummy))]
+pub enum ConstructionVersion {
+    /// The first version for constructing a UTXO
+    V0,
+}
+
+/// The sighash and enough metadata to piece together what happened.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct BitcoinSighash {
+    /// The transaction ID of the bitcoin transaction that sweeps funds
+    /// into and/or out of the signers' UTXO.
+    pub txid: BitcoinTxId,
+    /// The bitcoin chain tip when the sign request was submitted. This is
+    /// used to ensure that we do not sign for more than one transaction
+    /// containing inputs
+    pub chain_tip: BitcoinBlockHash,
+    /// The txid that created the output that is being spent.
+    pub prevout_txid: BitcoinTxId,
+    /// The index of the vout from the transaction that created this
+    /// output.
+    pub prevout_output_index: u32,
+    /// The sighash associated with the prevout.
+    pub sighash: TapSighash,
+    /// The type of prevout that we are dealing with.
+    pub prevout_type: TxPrevoutType,
+    /// The result of validation that was done on the input. For deposits,
+    /// this specifies whether validation succeeded and the first condition
+    /// that failed during validation. The signers' input is always valid,
+    /// since it is unconfirmed.
+    pub validation_result: InputValidationResult,
+    /// Whether the transaction is valid. A transaction is invalid if any
+    /// of the inputs or outputs failed validation.
+    pub is_valid_tx: bool,
+    /// Whether the signer will participate in a signing round for the
+    /// sighash.
+    pub will_sign: bool,
```

Comment:
If we check this information here instead of the `DepositSigner` model, should we also add a field like `was_signed` to indicate whether it has already been signed?

Reply:
I don't think we should check the `DepositSigner` model or associated table. Those just tell you whether you can sign or can accept the deposit request. Whether a signer will sign a specific sighash depends on the whether their inclusion is valid, and whether other deposits and withdrawals in the transaction are valid as well.